src/junglescout/async_client.py

```diff
@@ -0,0 +1,132 @@
+    ShareOfVoice,
+)
+from junglescout.session import AsyncSession
+
+
+class AsyncClient(BaseClient):
+    """The Jungle Scout API client.
+
+    This class is used to make requests to the Jungle Scout API. It provides methods to retrieve keyword data,
+    sales estimates, historical search volume, and share of voice data.
+    """
+
+    def __init__(
+        self,
+        api_key_name: str,
+        api_key: str,
+        api_type: ApiType = ApiType.JS,
+        marketplace: Optional[Marketplace] = None,
+    ):
+        """Initializes the Jungle Scout API client.
+
+        Args:
+            api_key_name: The name of the API key.
+            api_key: The API key.
+            api_type: The type of the API.
+            marketplace: The default marketplace to use for API requests.
+        """
+        super().__init__(api_key_name, api_key, api_type, marketplace)
+        headers = self._build_headers()
+        self.session = AsyncSession(headers)
+        self.session.login(api_key_name=api_key_name, api_key=api_key, api_type=api_type)
+        self.marketplace = marketplace
+
+    async def keywords_by_asin(
+        self,
+        asin: Union[str, List[str]],
+        include_variants: bool = True,
+        filter_options: Optional[FilterOptions] = None,
+        sort_option: Optional[Sort] = None,
+        marketplace: Optional[Marketplace] = None,
+        page_size: Optional[int] = None,
+        page: Optional[str] = None,
+    ) -> APIResponse[List[KeywordByASIN]]:
+        """Get keywords by ASIN.
+
+        Args:
+            asin: The ASIN (Amazon Standard Identification Number) of the product. This can be a list of
+                ASINs or a single ASIN.
+            include_variants: Include variants in the response.
+            filter_options: Filter options for the request.
+            sort_option: Sort option for the request.
+            marketplace: The Amazon marketplace.
+            page_size: The number of results to return per page.
+            page: The page to return.
+
+        Returns:
+            The response from the API.
+        """
+        params = KeywordByAsinParams(
+            marketplace=self._resolve_marketplace(marketplace), sort=sort_option, page=page, page_size=page_size
+        )
+
+        attributes = KeywordByAsinAttributes(
+            asin=asin, filter_options=filter_options, include_variants=include_variants
+        )
+
+        keyword_by_asin_request = KeywordByAsinRequest(params=params, attributes=attributes)
+
+        url = self.session.build_url(
+            "keywords", keyword_by_asin_request.type.value, params=keyword_by_asin_request.params
+        )
+
+        payload = keyword_by_asin_request.payload
+
+        response = await self.session.request("POST", url, json=payload)
+
+        if response.status_code == 200:
+            return APIResponse[List[KeywordByASIN]].model_validate(response.json())
+        else:
+            self._raise_for_status(response)
```

Comment:
There's going to be duplicate code across the `SyncClient` and the `AsyncClient` for each method. Consider abstracting generation of the `url` and `payload` to reduce duplication.

Reply:
I was thinking move the first part of the function to the base class, something like:

```suggestion
    async def keywords_by_asin(
        self,
        asin: Union[str, List[str]],
        include_variants: bool = True,
        filter_options: Optional[FilterOptions] = None,
        sort_option: Optional[Sort] = None,
        marketplace: Optional[Marketplace] = None,
        page_size: Optional[int] = None,
        page: Optional[str] = None,
    ) -> APIResponse[List[KeywordByASIN]]:
        """Get keywords by ASIN.

        Args:
            asin: The ASIN (Amazon Standard Identification Number) of the product. This can be a list of
                ASINs or a single ASIN.
            include_variants: Include variants in the response.
            filter_options: Filter options for the request.
            sort_option: Sort option for the request.
            marketplace: The Amazon marketplace.
            page_size: The number of results to return per page.
            page: The page to return.

        Returns:
            The response from the API.
        """
        url, payload = super().keywords_by_asin(...)
        response = await self.session.request("POST", url, json=payload)

        if response.status_code == 200:
            return APIResponse[List[KeywordByASIN]].model_validate(response.json())
        else:
            self._raise_for_status(response)
```

maybe the second part can also be abstracted

Reply:
Moving logic that produces the `url` / `payload` into the parent class is possible solution. If we were to do that, I'm thinking about an approach like this:

```python
class BaseClient(ABC):

    # define the public method in the parent class, to ensure the interface is the same
    @abstractmethod
    def keywords_by_asin(
        self,
        asin: Union[str, List[str]],
        include_variants: bool = True,
        filter_options: Optional[FilterOptions] = None,
        sort_option: Optional[Sort] = None,
        marketplace: Optional[Marketplace] = None,
        page_size: Optional[int] = None,
        page: Optional[str] = None,
    ) -> Union[APIResponse[List[KeywordByASIN]], Awaitable[APIResponse[List[KeywordByASIN]]]]:
        pass

    def _prepare_request_keywords_by_asin(
        self, asin, include_variants, filter_options, sort_option, marketplace, page_size, page
    ) -> Tuple[str, dict]:
        params = KeywordByAsinParams(
            marketplace=self._resolve_marketplace(marketplace), sort=sort_option, page=page, page_size=page_size
        )

        attributes = KeywordByAsinAttributes(
            asin=asin, filter_options=filter_options, include_variants=include_variants
        )

        keyword_by_asin_request = KeywordByAsinRequest(params=params, attributes=attributes)

        url = self._build_url("keywords", keyword_by_asin_request.type.value, params=keyword_by_asin_request.params)
        payload = self._build_payload(keyword_by_asin_request)

        return url, payload
```

Then in the subclass...

```python
class AsyncClient(BaseClient):
    async def keywords_by_asin(
        self,
        asin: Union[str, List[str]],
        include_variants: bool = True,
        filter_options: Optional[FilterOptions] = None,
        sort_option: Optional[Sort] = None,
        marketplace: Optional[Marketplace] = None,
        page_size: Optional[int] = None,
        page: Optional[str] = None,
    ) -> APIResponse[List[KeywordByASIN]]:
        url, payload = self._prepare_request_keywords_by_asin(
            asin, include_variants, filter_options, sort_option, marketplace, page_size, page
        )
        response = await self.session.request("POST", url, json=payload)
        if response.status_code == 200:
            return APIResponse[List[KeywordByASIN]].model_validate(response.json())
        else:
            self._raise_for_status(response)
```

Another option would be for us to extend the `KeywordByAsinRequest` model to include the repetitive boilerplate. Here's an example...

```python
# junglescout/models/requests/base_request.py

class BaseRequest(ABC, Generic[ParamsType, AttributesType]):

    # adding a `url` property 
    @property
    @abstractmethod
    def url(self) -> str:
        pass

    # pass in the session instance
    def __init__(self, params: ParamsType, attributes: AttributesType, session: BaseSession):
        # store additional data in each request instance 
        self.params_model_instance = params
        self.attributes_model_instance = attributes
        self.session = session

# junglescout/models/requests/keyword_by_asin_request.py

class KeywordByAsinRequest(BaseRequest[KeywordByAsinParams, KeywordByAsinAttributes]):
    # subclasses define the URL they need to use
    @property
    def url(self) -> str:
        return self.session.build_url("keywords", self.type.value, params=self.params)

    # introduce a factory method that handles creation of the `KeywordByAsinParams` and `KeywordByAsinAttributes` models
    @classmethod
    def from_args(cls, args: "KeywordByAsinRequestArgs", session: BaseSession) -> "KeywordByAsinRequest":
        params = KeywordByAsinParams(
            marketplace=args.marketplace, sort=args.sort_option, page=args.page, page_size=args.page
        )
        attributes = KeywordByAsinAttributes(
            asin=args.asin,
            filter_options=args.filter_options,
            include_variants=args.include_variants,
        )
        return cls(params=params, attributes=attributes, session=session)

# introduce an `Args` model that makes it easier to pass data around
class KeywordByAsinRequestArgs(BaseModel):
    asin: Union[str, List[str]]
    include_variants: bool = True
    filter_options: Optional[FilterOptions] = None
    sort_option: Optional[Sort] = None
    marketplace: Optional[Marketplace] = None
    page_size: Optional[int] = None
    page: Optional[str] = None

# junglescout/async_client.py

class AsyncClient(BaseClient):
    async def keywords_by_asin(
        self,
        asin: Union[str, List[str]],
        include_variants: bool = True,
        filter_options: Optional[FilterOptions] = None,
        sort_option: Optional[Sort] = None,
        marketplace: Optional[Marketplace] = None,
        page_size: Optional[int] = None,
        page: Optional[str] = None,
    ) -> APIResponse[List[KeywordByASIN]]:
        args = KeywordByAsinRequestArgs(
            asin=asin,
            include_variants=include_variants,
            filter_options=filter_options,
            sort_option=sort_option,
            marketplace=self._resolve_marketplace(marketplace),
            page_size=page_size,
            page=page,
        )
        request_instance = KeywordByAsinRequest.from_args(args, self.session)
        response = await self.session.request("POST", request_instance.url, json=request_instance.payload)
        if response.status_code == 200:
            return APIResponse[List[KeywordByASIN]].model_validate(response.json())
        else:
            self._raise_for_status(response)
```

I think I prefer approach #2. What do you think?

Reply:
I like the second approach more (Extending KeywordByAsinRequest). But for simplicity and moving forward faster with this PR maybe the first one will be better. For me, time is more important in this case, maybe the second approach can come in a later version. For you which option do you prefer?